scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3WorkerCQL.java

```diff
@@ -130,7 +133,20 @@ public void onFailure(Throwable t) {
                     // it wouldn't be nice if we had to update `schema` on each query/page (expensive)
                     // See Scylla issue #7824.
                 }
-                fut.complete(Optional.of(new Driver3RawChange(row, schema)));
+                Driver3RawChange newChange = new Driver3RawChange(row, schema);
+
+                // lastChangeId determines the point from which we should
+                // start reading within a window. In this implementation
+                // we simply read the entire window and skip rows that
+                // were before lastChangeId.
+                //
+                // If lastChangeId is Optional.empty(), then we read
+                // the entire window.
+                if (!lastChangeId.isPresent() || newChange.getId().compareTo(lastChangeId.get()) > 0) {
```

Comment:
It would probably be better to adjust the select to fetch only the required changes but that can probably be done in another PR.